### ISSUE TITLE
Wrong error message with if statement in a function returning non-()

### ISSUE THREAD

**Issue by pengowen123** [rust-lang/rust#39236]
I have this function:
```
fn foo() -> i32 {  
    if true {}  
}
```
When compiled on stable (1.14), I get this error:

    error[E0308]: mismatched types
      --> <anon>:2:5
      |
    2 |     if true {}
      |     ^^^^^^^^^^ expected i32, found ()
      |
       = note: expected type `i32`
       = note:    found type `()`

When compiled on 1.16.0-nightly (f0b420759 2017-01-19), an additional error appears:

    error[E0317]: if may be missing an else clause
     --> <anon>:2:5
      |
    2 |     if true {}
      |     ^^^^^^^^^^ expected (), found i32
      |
      = note: expected type `()`
      = note:    found type `i32`

A missing else clause isn't the problem. Also, it says it expected `()`, but found `i32`, which is confusing because there are no numbers anywhere in the code.

**Comment by Mark-Simulacrum** [comment]
I get neither of those errors with the code example you gave. Given playground's stable and nightly, I see the below output. Can you verify that the code example you gave is the one you compiled?

```rust
fn foo() {  
    if true {}  
}

fn main() {}
```

produces

```
rustc 1.16.0-nightly (f0b420759 2017-01-19)
warning: function is never used: `foo`, #[warn(dead_code)] on by default
 --> <anon>:1:1
  |
1 |   fn foo() {  
  |  _^ starting here...
2 | |     if true {}  
3 | | }
  | |_^ ...ending here
```

and

```
rustc 1.14.0 (e8a012324 2016-12-16)
warning: function is never used: `foo`, #[warn(dead_code)] on by default
 --> <anon>:1:1
  |
1 | fn foo() {  
  | ^
```

**Comment by pengowen123** [comment]
`foo` should return something that is not `()`, such as i32.

**Comment by Mark-Simulacrum** [comment]
This is what I see on nightly (with `-> i32`):
```
rustc 1.16.0-nightly (f0b420759 2017-01-19)
error[E0308]: mismatched types
 --> <anon>:2:13
  |
2 |     if true {}  
  |             ^^ expected i32, found ()
  |
  = note: expected type `i32`
  = note:    found type `()`

error[E0317]: if may be missing an else clause
 --> <anon>:2:5
  |
2 |     if true {}  
  |     ^^^^^^^^^^ expected (), found i32
  |
  = note: expected type `()`
  = note:    found type `i32`

error: aborting due to 2 previous errors
```

The situation could be improved, but the two errors are sort of communicating the truth. The first one states that the if is expected to return an i32, which is correct, while the second says that the if expression's result should be `()` since the first and only branch returns that, but the "receiver" expected a `i32`, which is incorrect.

I think we shouldn't bring both errors, but I'm not sure which is better in this case. Probably the mismatched types. However, I feel like this specific scenario may be hard to detect.

**Comment by pengowen123** [comment]
Maybe the missing else clause error should not be given if the if block doesn't return the correct type (in that situation adding an else block won't fix anything).

**Comment by Mark-Simulacrum** [comment]
We no longer give the missing else clause error. As such, I'm going to go ahead and close.

```
rustc 1.19.0-dev (5dfcd85fd 2017-05-19)
error[E0308]: mismatched types
 --> test.rs:2:13
  |
2 |     if true {}
  |             ^^ expected i32, found ()
  |
  = note: expected type `i32`
             found type `()`

error: aborting due to previous error
```
